File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
'''
N = nums.length

T: O(N)
S: O(1)

'''

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        total = 0
        zeroCount = 0
        
        for num in nums:
            if num != 0:
                if total == 0:
                    total = 1
                
                total *= num
            else:
                zeroCount += 1
                
                
        if zeroCount > 1:
            return [0] * len(nums)
        elif zeroCount == 1:
            for i in range(len(nums)):
                if nums[i] == 0:
                    nums[i] = total
                else:
                    nums[i] =  0
        else:
            for i in range(len(nums)):
                nums[i] = total // nums[i]

            
        return nums
            

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    /*
        // Algorithm & DS: prefixSum
        // Time: O(n), Extra Space: O(1)
    */
    public int[] productExceptSelf(int[] nums) {
        int[] res = new int[nums.length];
        int prefixSum = 1;
        int zeroIdx = -1;

        for(int i = 0; i < nums.length; i++){
            if(nums[i] == 0){
                if (zeroIdx >= 0){
                    return res;
                }
                else {
                    zeroIdx = i;
                }
            }
            else {
                prefixSum *= nums[i];
            }
        }

        if(zeroIdx >= 0){
            res[zeroIdx] = prefixSum;
        }
        else {
            for(int i = 0; i < nums.length; i++){
                res[i] = prefixSum / nums[i];
            }
        }

        return res;
    }
}